app/mqtt.mjs

```diff
@@ -1,15 +1,264 @@
+        // defaults for MQTT switches in Home Assistant
+        topicMap[topicName] = state ? 'ON' : 'OFF'
+    }
+
+    // Publish each reading
     const readings = await getReadings(modbusClient)
-    let topicMap = {}
 
     for (const [reading, value] of Object.entries(readings)) {
         const topicName = `${TOPIC_PREFIX}/readings/${reading}`
-        topicMap[topicName] = value
+        topicMap[topicName] = JSON.stringify(value)
+    }
+
+    // Publish each setting
+    const settings = await getSettings(modbusClient)
+
+    for (const [setting, value] of Object.entries(settings)) {
+        const topicName = `${TOPIC_PREFIX}/settings/${setting}`
+        topicMap[topicName] = JSON.stringify(value)
+    }
+
+    const publishPromises = []
+
+    for (const [topic, value] of Object.entries(topicMap)) {
+        publishPromises.push(mqttClient.publish(topic, value))
+    }
+
+    await Promise.all(publishPromises)
+}
+
+export const subscribeToSettingChanges = async (modbusClient, mqttClient) => {
+    // Subscribe to both settings and mode changes
+    const topicNames = [
+        `${TOPIC_PREFIX}/mode/+/set`,
+        `${TOPIC_PREFIX}/settings/+/set`,
+    ]
+
+    for (const topicName of topicNames) {
+        console.log(`Subscribing to topic(s) ${topicName}`)
+
+        await mqttClient.subscribe(topicName)
+    }
+}
+
+export const handleMessage = async (modbusClient, mqttClient, topicName, payload) => {
+    // Payload looks like a string when logged, but any comparison with === will equal false unless we convert the
+    // Buffer to a string
+    const payloadString = payload.toString()
+
+    console.log(`Received ${payloadString} on topic ${topicName}`)
+
+    // Handle settings updates
+    if (topicName.startsWith('eda/settings/') && topicName.endsWith('/set')) {
+        const settingName = topicName.substring('eda/settings/'.length, topicName.lastIndexOf('/'))
+
+        console.log(`Updating setting ${settingName} to ${payloadString}`)
+
+        await setSetting(modbusClient, settingName, payloadString)
+    } else if (topicName.startsWith('eda/mode/') && topicName.endsWith('/set')) {
+        const mode = topicName.substring('eda/mode/'.length, topicName.lastIndexOf('/'))
+
+        console.log(`Updating mode ${mode} to ${payloadString}`)
+
+        await setFlag(modbusClient, mode, payloadString === 'ON')
+    }
+
+    // Publish all values again for state changes to "take"
+    await publishValues(modbusClient, mqttClient)
+}
+
+export const configureMqttDiscovery = async (modbusClient, mqttClient) => {
+    const modbusDeviceInformation = await getDeviceInformation(modbusClient)
+    const familyType = modbusDeviceInformation.familyType
+    const serialNumber = modbusDeviceInformation.serialNumber
+    const softwareVersion = modbusDeviceInformation.softwareVersion
+    const deviceIdentifier = `Enervent-${familyType}-${serialNumber}-${softwareVersion}`;
```

Comment:
I now that these devices don´t get software updates too often, but maybe version could be left out from here.

Reply:
Yeah, I think I'll simplify this to just contain the family type and "Enervent". The serial number is always 0 as well, unless someone has manually updated the register value, so it's not very useful.

Reply:
Split into https://github.com/Jalle19/eda-modbus-bridge/issues/14